TechnicalDocuments/Proposals/IdentifiersForUIAutomation.md

```diff
@@ -0,0 +1,203 @@
+        return prefix + automationName
+    }
+}
+```
+
+and this particular composite component called `prettyCard`:
+
+```swift
+extension DesignLibrary.ComponentsLibrary {
+    public func prettyCard() -> AnyRenderable {
+        return [
+            UI.Label(/* ... */, automationName: String(AutomationName.prettyCardNotice)),
+            UI.Button(/* ... */, automationName: String(AutomationName.prettyCardSubmit))
+        ].stack(axis: .vertical)
+    }
+}
+
+extension DesignLibrary.AutomationName {
+    public enum PrettyCard: String {
+        case notice
+        case submit
+    }
+}
+```
+
+When we render `prettyCard`:
+
+```
+render(
+    Box(
+        sections: [
+            Section(id: .activity, items: [
+                Node(id: .card, component:
+                    prettyCard()
+                )
+            ])
+        ]
+    )
+)
+```
+
+It should result in two accessibility identifiers:
+
+* `activity/card/notice` applied to the label; and
+* `activity/card/submit` applied to the button.
+
+Notice that automation names for the relevant subcomponents of `prettyCard` has been published in the `DesignLibrary.AutomationName` umbrella. This facilitates the next suggestion of this proposal...
+
+### Working with UI automation tests
+
+Now that we have a systematic approach of defining and publishing information about accessibility identifiers, we can use this to reliably generate accessibility identifiers in UI automation tests!
+
+The proposal suggests that the UI automation test target should link with all UI frameworks. This allows the automation code to have access to `Renderer`s and any published `AutomationName` enums, and can therefore construct accessibility identifiers. Changes made to any section ID, item ID or automation name enums would either be automatically picked up by the automation code, or lead to compile-time breakage depending on the circusmtances.
+
+`BaseScreen` shall be extended to provide a conveinence:
+```swift
+extension BaseScreen {
+    subscript<Renderer: BoxRenderer>(
+        _ renderer: Renderer.Type,
+        _ section: Renderer.SectionID,
+        _ item: Renderer.ItemID
+    ) -> String {
+        return "\(section)/\(item)/"
+    }
+
+    subscript<Renderer: BoxRenderer, AutomationName: RawRepresentable>(
+        _ renderer: Renderer.Type,
+        _ section: Renderer.SectionID,
+        _ item: Renderer.ItemID,
+        _ automationName: AutomationName
+    ) -> String where AutomationName.RawValue == String {
+        return "\(section)/\(item)/" + automationName.rawValue
+    }
+}
+```
+
+which the automation code can rely on. For example:
+
+```swift
+import BabylonHealthManagementUI
+
+extension BaseScreen {
+    // NOTE: alias in the UI testing target to help reduce verbosity.
+    typealias Name = DesignLibrary.AutomationName
+}
+
+class FeedScreen: BaseScreen {
+    func tapSubmit() {
+        let button = app[FeedRenderer.self, .activity, .card, Name.PrettyCard.submit]
```

Comment:
the subscript is defined on BaseScreen, but applied here in the `app`, which is `XCUIElement`. We shouldn't be forced to make queries in the app object, as it takes more time to perform query. At least it should be still possible to do `app.buttons[...]`, ideally there shouldn't be limitation in how elements are accessed.

Reply:
I'd suggest to try implement API like this:

```swift
// root table view should have id based on type of renderer
let table = app.tables[FeedRenderer.self]

//most specific query, search for button in all cells, use section and node id as part of cell identifier
screen.cells[.activity, .card].buttons[.submit]

// same as 
table.cells["activity_card"].buttons["submit"]

// or least specific query, search among all buttons using all ids to make final identifier
screen.buttons[.activity, .card, .submit]

// same as
table.buttons["activity_card_submit"]
```

I suppose we wouldn't generally have table views in table view (and in this case inner table view should be queried in two steps any way), so subscript with maximum 3 parameters (section, node, component) should be enough.

Reply:
Agreed we should not be doing app queries in the app

Reply:
We can move the extension to `XCUIElementQuery`.

Reply:
Re: Two-level queries
Proposal has been updated to state why this doesn't work at all with the current Accessibility mechanism.